scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -407,107 +417,232 @@
+         :upgrade-backfill upgrade-backfill
+         :backfill-jobs backfill-jobs
+         :matched-head? matched-head?}))))
+
+(defn handle-resource-offers!
   "Gets a list of offers from mesos. Decides what to do with them all--they should all
    be accepted or rejected at the end of the function."
-  [conn driver fid pending-jobs offer]
-  (try
+  [conn driver ^TaskScheduler fenzo fid pending-jobs num-considerable offers-chan offers]
+  (doseq [offer offers]
     (histograms/update!
-     offer-size-cpus
-     (get-in offer [:resources :cpus]))
+      offer-size-cpus
+      (get-in offer [:resources :cpus] 0))
     (histograms/update!
-     offer-size-mem
-     (get-in offer [:resources :mem]))
-    (catch Throwable t
-      (log/warn t "Error in updating offer metrics:" (ex-data t))))
-
-  (timers/time!
-    handle-resource-offer!-duration
-    (try
-      (loop [] ;; This loop is for compare-and-set! below
-        (let [scheduler-contents @pending-jobs
-              db (db conn)
-              considerable (filter (fn [job]
-                                     (and (try
-                                            (d/invoke db :job/allowed-to-start? db (:db/id job))
-                                            true
-                                            (catch clojure.lang.ExceptionInfo e
-                                              false))
-                                          ;; ensure it always tries new hosts
-                                          (not-any? #(= % (:hostname offer))
-                                                    (->> (:job/instance job)
-                                                         (filter #(false? (:instance/preempted? %)))
-                                                         (map :instance/hostname)))))
-                                   scheduler-contents)
-              matches (match-offer-to-schedule considerable offer)
-              ;; We know that matches always form a prefix of the scheduler's contents
-              new-scheduler-contents (remove (set matches) scheduler-contents)
-              first-considerable-resources (-> considerable first util/job-ent->resources)
-              match-resource-requirements (util/sum-resources-of-jobs matches)]
-          (reset! front-of-job-queue-mem-atom
-                  (or (:mem first-considerable-resources) 0))
-          (reset! front-of-job-queue-cpus-atom
-                  (or (:cpus first-considerable-resources) 0))
-          (cond
-            (empty? matches) (decline-offers driver (:ids offer))
-            (not (compare-and-set! pending-jobs scheduler-contents new-scheduler-contents))
-            (do (log/info "Pending job atom contention encountered")
+      offer-size-mem
+      (get-in offer [:resources :mem] 0)))
+  (log/debug "invoked handle-resource-offers!")
+  (let [offer-stash (atom nil)] ;; This is a way to ensure we never lose offers fenzo assigned if an errors occures in the middle of processing
```

Comment:
Maybe put a comment:

If we fail to launch some of the tasks, we will put all the offers - used and unused - onto offer-stash. Those used offers will be feed back to frenzo again for scheduling, which will probably cause task lost next time we use it.